lang/derive/space/src/lib.rs

```diff
@@ -0,0 +1,173 @@
+        quote!(anchor_lang::__private::max(#item, #next_item))
+    } else {
+        quote!(0)
+    }
+}
+
+fn len_from_type(ty: Type, attrs: &mut Option<IntoIter<LitInt>>) -> TokenStream2 {
+    match ty {
+        Type::Array(TypeArray { elem, len, .. }) => {
+            let array_len = len.to_token_stream();
+            let type_len = len_from_type(*elem, attrs);
+            quote!((#array_len * #type_len))
+        }
+        Type::Path(path) => {
+            let path_segment = path.path.segments.last().unwrap();
+            let ident = &path_segment.ident;
+            let type_name = ident.to_string();
+            let first_ty = get_first_ty_arg(&path_segment.arguments);
+
+            match type_name.as_str() {
+                "i8" | "u8" | "bool" => quote!(1),
+                "i16" | "u16" => quote!(2),
+                "i32" | "u32" | "f32" => quote!(4),
+                "i64" | "u64" | "f64" => quote!(8),
+                "i128" | "u128" => quote!(16),
+                "String" => {
+                    let max_len = get_next_arg(ident, attrs);
+                    quote!((4 + #max_len))
+                }
+                "Pubkey" => quote!(32),
+                "Option" => {
+                    if let Some(ty) = first_ty {
+                        let type_len = len_from_type(ty, attrs);
+
+                        quote!((1 + #type_len))
+                    } else {
+                        quote_spanned!(ident.span() => compile_error!("Invalid argument in Vec"))
+                    }
+                }
+                "Vec" => {
+                    if let Some(ty) = first_ty {
+                        let max_len = get_next_arg(ident, attrs);
+                        let type_len = len_from_type(ty, attrs);
+
+                        quote!((4 + #type_len * #max_len))
+                    } else {
+                        quote_spanned!(ident.span() => compile_error!("Invalid argument in Vec"))
+                    }
+                }
+                _ => {
+                    let ty = &path_segment.ident;
+                    quote!(<#ty as anchor_lang::Space>::INIT_SPACE)
+                }
+            }
+        }
+        _ => panic!("Type {:?} is not supported", ty),
+    }
+}
+
+fn get_first_ty_arg(args: &PathArguments) -> Option<Type> {
+    match args {
+        PathArguments::AngleBracketed(bracket) => bracket.args.iter().find_map(|el| match el {
+            GenericArgument::Type(ty) => Some(ty.to_owned()),
+            _ => None,
+        }),
+        _ => None,
+    }
+}
+
+fn get_max_len_args(attributes: &[Attribute]) -> Option<IntoIter<LitInt>> {
+    attributes
+        .iter()
+        .find(|a| a.path.is_ident("max_len"))
+        .and_then(|a| {
+            a.parse_args_with(Punctuated::<LitInt, Token![,]>::parse_terminated)
+                .ok()
+        })
+        .map(|p| p.into_iter())
+}
+
+fn get_next_arg(ident: &Ident, args: &mut Option<IntoIter<LitInt>>) -> TokenStream2 {
```

Comment:
While I was initially confused by this function, i see now that it exists to support nested types of multiple lengths. A small explanation of how to use `max_len` should probably be added to the doc string. e.g. `For variable length types use `max_len(x)` attribute above the field. Nested types with variables length e.g. Vec<Vec<u8>> can be specified using max_len(2, 4)`

Reply:
Is it ok for the doc now?

Reply:
looks good!